tools/distrib/python/xds_protos/build.py

```diff
@@ -0,0 +1,135 @@
+#! /usr/bin/env python3
+# Copyright 2021 The gRPC Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Builds the content of xds-protos package"""
+
+import os
+import pkg_resources
+from grpc_tools import protoc
+
+# We might not want to compile all the protos
+EXCLUDE_PROTO_PACKAGES_LIST = [
+    # Requires extra dependency to Prometheus protos
+    'envoy/service/metrics/v2',
+    'envoy/service/metrics/v3',
+    'envoy/service/metrics/v4alpha',
+]
+
+# Compute the pathes
+WORK_DIR = os.path.dirname(os.path.abspath(__file__))
+GRPC_ROOT = os.path.abspath(os.path.join(WORK_DIR, '..', '..', '..', '..'))
+XDS_PROTO_ROOT = os.path.join(GRPC_ROOT, 'third_party', 'envoy-api')
+UDPA_PROTO_ROOT = os.path.join(GRPC_ROOT, 'third_party', 'udpa')
+GOOGLEAPIS_ROOT = os.path.join(GRPC_ROOT, 'third_party', 'googleapis')
+VALIDATE_ROOT = os.path.join(GRPC_ROOT, 'third_party', 'protoc-gen-validate')
+OPENCENSUS_PROTO_ROOT = os.path.join(GRPC_ROOT, 'third_party',
+                                     'opencensus-proto', 'src')
+WELL_KNOWN_PROTOS_INCLUDE = pkg_resources.resource_filename(
+    'grpc_tools', '_proto')
+OUTPUT_PATH = WORK_DIR
+
+# Prepare the test file generation
+TEST_FILE_NAME = 'generated_file_import_test.py'
+TEST_IMPORTS = []
+
+
+def add_test_import(proto_package_path: str,
+                    file_name: str,
+                    service: bool = False):
+    TEST_IMPORTS.append("from %s import %s\n" % (proto_package_path.replace(
+        '/', '.'), file_name.replace('.proto', '_pb2')))
+    if service:
+        TEST_IMPORTS.append("from %s import %s\n" % (proto_package_path.replace(
+            '/', '.'), file_name.replace('.proto', '_pb2_grpc')))
+
+
+# Prepare Protoc command
+COMPILE_PROTO_ONLY = [
+    'grpc_tools.protoc',
+    '--proto_path={}'.format(XDS_PROTO_ROOT),
+    '--proto_path={}'.format(UDPA_PROTO_ROOT),
+    '--proto_path={}'.format(GOOGLEAPIS_ROOT),
+    '--proto_path={}'.format(VALIDATE_ROOT),
+    '--proto_path={}'.format(WELL_KNOWN_PROTOS_INCLUDE),
+    '--proto_path={}'.format(OPENCENSUS_PROTO_ROOT),
+    '--python_out={}'.format(OUTPUT_PATH),
+]
+COMPILE_BOTH = COMPILE_PROTO_ONLY + ['--grpc_python_out={}'.format(OUTPUT_PATH)]
+
+
+def has_grpc_service(proto_package_path: str) -> bool:
+    return proto_package_path.startswith('envoy/service')
```

Comment:
Is this assumption guaranteed going forward? What's the behavior when you use `--grpc_python_out` on a file with no services? If it emits an empty file or doesn't emit any file at all, you might just want to unconditionally supply this flag.

Reply:
Our ProtoBuf plugin will create a file if the `--grpc_python_out` flag is given. The file will have a line to import the generated `_pb.py` of that proto. When trying to import generated proto modules, the extra `_pb_grpc` will appear in the code hinter (I'm using VSCode). So, I added the logic to only generate service when we are certain there is a gRPC service.

So far, I think Envoy protos are doing a good job to keep services under `envoy/service`. If there are any exceptions, I can update the code. Or, do you think we should use regex to read the proto file?

Reply:
> So far, I think Envoy protos are doing a good job to keep services under envoy/service

If there is a problem, it should happen infrequently enough that manually fixes should suffice.